File: Exercism/fsharp/interest-is-interesting/InterestIsInteresting.fs
module InterestIsInteresting

let interestRate =
    function
    | b when b < 0m -> 3.213f
    | b when b < 1_000m -> 0.5f
    | b when b < 5_000m -> 1.621f
    | _ -> 2.475f

let interest amount = (amount |> interestRate |> decimal) * amount / 100m

let annualBalanceUpdate balance = (balance |> interest) + balance

let amountToDonate (balance: decimal) (taxFreePercentage: float) =
    if balance < 0m then 0
    else balance * (taxFreePercentage |> decimal) / 50m |> int

File: Exercism/csharp/interest-is-interesting/InterestIsInteresting.cs
static class SavingsAccount
{
  public static float InterestRate(decimal balance) => balance switch
  {
    < 0m => 3.213f,
    < 1_000m => 0.5f,
    < 5_000m => 1.621f,
    _ => 2.475f
  };

  public static decimal Interest(decimal amount) => (decimal)InterestRate(amount) * amount / 100m;
  public static decimal AnnualBalanceUpdate(decimal balance) => balance + Interest(balance);
  public static int YearsBeforeDesiredBalance(decimal balance, decimal targetBalance) => YearsBeforeDesiredBalance(balance, targetBalance, 1);
  private static int YearsBeforeDesiredBalance(decimal balance, decimal targetBalance, int yearsElapsed) =>
    AnnualBalanceUpdate(balance) < targetBalance
    ? YearsBeforeDesiredBalance(AnnualBalanceUpdate(balance), targetBalance, yearsElapsed + 1)
    : yearsElapsed;
}